esmlab/statistics.py

```diff
@@ -5,357 +5,389 @@
-           By default weighted correlation is applied over all dimensions.
+        Dimension(s) over which to apply correlation.
+    weights : DataArray
+        weights to apply. Shape must be broadcastable to shape of data.
 
-    apply_nan_mask : bool, default: True
+    return_p : bool, default: True
+        If True, compute and return the p-value(s) associated with the
+        correlation.
+
+    return_p : bool, default: True
+        If True, compute and return the p-value(s) associated with the
+        correlation.
 
     Returns
     -------
+    reduced : Dataset or DataArray
+        New Dataset/DataArray with correlation applied to x, y and the indicated
+        dimension(s) removed.
 
-    weighted_correlation : float
-              If `weights` is None, returns correlation using equal weights for all data points.
-
-
+    pval : float
+              If `return_p` is True, returns the p value from a two-tailed t-test.
     """
-    if weights is None:
-        warn('Computing weighted correlation using equal weights for all data points')
 
-    weights, op_over_dims = _get_weights_and_dims(
-        x, weights=weights, dim=dim, apply_nan_mask=apply_nan_mask
-    )
+    valid_values = x.notnull() & y.notnull()
+    x = x.where(valid_values)
+    y = y.where(valid_values)
+    numerator = weighted_cov(x, y, dim, weights)
+    denominator = np.sqrt(weighted_cov(x, x, dim, weights) * weighted_cov(y, y, dim, weights))
+    corr_xy = numerator / denominator
 
-    # If the mask is applied in previous operation,
-    # disable it for subseqent operations to speed up computation
-    if apply_nan_mask:
-        apply_nan_mask_flag = False
+    if return_p:
+        p = compute_corr_significance(corr_xy, len(x))
+        return corr_xy, p
     else:
-        apply_nan_mask_flag = True
-
-    numerator = weighted_cov(
-        x=x, y=y, weights=weights, dim=op_over_dims, apply_nan_mask=apply_nan_mask_flag
-    )
-    denominator = np.sqrt(
-        weighted_cov(x, x, weights=weights, dim=op_over_dims, apply_nan_mask=apply_nan_mask_flag)
-        * weighted_cov(y, y, weights=weights, dim=op_over_dims, apply_nan_mask=apply_nan_mask_flag)
-    )
-    corr_xy = numerator / denominator
-    return corr_xy
+        return corr_xy
+
+
+@esmlab_xr_set_options(arithmetic_join='exact', keep_attrs=True)
+def compute_corr_significance(r, N):
+    """ Compute statistical significance for a pearson correlation between
+        two xarray objects.
+
+    Parameters
+    ----------
+    r : `xarray.DataArray` object
+        correlation coefficient between two time series.
+
+    N : int
+        length of time series being correlated.
+
+    Returns
+    -------
+    pval : float
+        p value for pearson correlation.
+
+    """
+    df = N - 2
+    t_squared = r**2 * (df / ((1.0 - r) * (1.0 + r)))
+    pval = special.betainc(0.5 * df, 0.5,
+                           np.fmin(np.asarray(df / (df + t_squared)), 1.0))
```

Comment:
@bradyrx, can we avoid casting this array to NumPy, i.e can this function work on both NumPy and Dask arrays? In case the DataArray is backed by distributed dask arrays, casting to NumPy may lead to some issues such as Memory Error, etc..

Reply:
Yes -- didn't realize `np.fmin` would replace values without issue in a DataArray. Fixed.